From 4d093781a04b3a83992d15c56e08d10c915c57b9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Robert=20L=C3=B6hning?= <robert.loehning@qt.io>
Date: Fri, 17 Oct 2025 11:38:17 +0200
Subject: [PATCH 430/553] Avoid float rounding errors in QDashStroker

Calculations which should have a result of zero can lead to slightly
different results when using floating point values. This patch fixes a
case which could trigger an assert although all values passed to the
public API were harmless.

Credit to OSS-Fuzz for finding an SVG file which triggered this as issue
429123947. I derived the test from the values in the file.

Pick-to: 6.8
Change-Id: I06d6f0f0bf6beb758f54423e16e9c5653ed8edf1
Reviewed-by: Eirik Aavitsland <eirik.aavitsland@qt.io>
(cherry picked from commit 697467451973a21ed3c8ed31e6d15c1abac85479)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/gui/painting/qstroker.cpp                 |  6 +++--
 .../gui/painting/qpainter/tst_qpainter.cpp    | 26 +++++++++++++++++++
 2 files changed, 30 insertions(+), 2 deletions(-)

diff --git a/src/gui/painting/qstroker.cpp b/src/gui/painting/qstroker.cpp
index 79799ca2ece..0d435c95048 100644
--- a/src/gui/painting/qstroker.cpp
+++ b/src/gui/painting/qstroker.cpp
@@ -1154,7 +1154,8 @@ void QDashStroker::processCurrentSubpath()
             elen -= std::floor(elen * invSumLength) * sumLength;
             // Update dash offset.
             while (!done) {
-                qreal dpos = pos + dashes[idash] - doffset - estart;
+                // parentheses to avoid float rounding issues: qreal(4) + 0.1 - 0.1 - 4 < 0
+                qreal dpos = (pos + dashes[idash]) - (doffset + estart);
 
                 Q_ASSERT(dpos >= 0);
 
@@ -1189,7 +1190,8 @@ void QDashStroker::processCurrentSubpath()
 
             bool has_offset = doffset > 0;
             bool evenDash = (idash & 1) == 0;
-            qreal dpos = pos + dashes[idash] - doffset - estart;
+            // parentheses to avoid float rounding issues: qreal(4) + 0.1 - 0.1 - 4 < 0
+            qreal dpos = (pos + dashes[idash]) - (doffset + estart);
 
             Q_ASSERT(dpos >= 0);
 
diff --git a/tests/auto/gui/painting/qpainter/tst_qpainter.cpp b/tests/auto/gui/painting/qpainter/tst_qpainter.cpp
index f0802e0150f..272201a0dcc 100644
--- a/tests/auto/gui/painting/qpainter/tst_qpainter.cpp
+++ b/tests/auto/gui/painting/qpainter/tst_qpainter.cpp
@@ -291,6 +291,8 @@ private slots:
     void alphaBlitToNonAlphaFormats_data();
     void alphaBlitToNonAlphaFormats();
 
+    void floatRounding();
+
 private:
     void fillData();
     void setPenColor(QPainter& p);
@@ -5703,6 +5705,30 @@ void tst_QPainter::alphaBlitToNonAlphaFormats()
     }
 }
 
+void tst_QPainter::floatRounding()
+{
+    // oss-fuzz issue 429123947
+    // The following triggered an assert in QDashStroker::processCurrentSubpath(): "dpos >= 0"
+    // when it expected the calculation's result to be zero but it was actually smaller:
+    // qreal(4) + qreal(0.1) - qreal(0.1) - qreal(4)
+    // actual result: -4.440892098500626e-16
+    QImage img(5, 5, QImage::Format_RGB888);
+    QPainter p(&img);
+
+    QList<qreal> pattern {0.1, 0.3, 0.1, 0.1, 0.3, 0.1};
+    QPainterPathStroker stroker;
+    stroker.setDashPattern(pattern);
+
+    QPainterPath pp;
+    pp.moveTo(4.0, 0.0);
+    pp.lineTo(0.1, 0.0);
+    pp.lineTo(0.0, 0.0);
+    pp.lineTo(0.0, 5.0);
+
+    QPolygonF poly = stroker.createStroke(pp).toFillPolygon();
+    p.drawPolygon(poly);
+}
+
 QTEST_MAIN(tst_QPainter)
 
 #include "tst_qpainter.moc"
-- 
2.51.2

